src/google/protobuf/compiler/csharp/csharp_message.cc

```diff
@@ -489,8 +500,12 @@ void MessageGenerator::GenerateFrameworkMethods(io::Printer* printer) {
         generator->WriteHash(printer);
     }
     for (int i = 0; i < descriptor_->oneof_decl_count(); i++) {
-        printer->Print("hash ^= (int) $name$Case_;\n",
-            "name", UnderscoresToCamelCase(descriptor_->oneof_decl(i)->name(), false));
+      // Skip synthetic oneofs (optional proto3 fields)
```

Comment:
You can now iterate up to `descriptor_->real_oneof_decl_count()`, to avoid the need for a test inside the loop. Ditto for other places in the code that are iterating over oneofs.

Reply:
Aha - that'll make things *much* simpler, thanks. (Not yet implemented, but will do so before I next push the change, so I'll resolve now.)